### ISSUE TITLE
[template/no-non-null-assertion] Rule complains about necessary assertion

### ISSUE THREAD

**Issue by rdamazio** [angular-eslint/angular-eslint#2924]
**Description and reproduction of the issue**

YOUR LINK HERE: https://stackblitz.com/edit/angular-eslint-repros-rft8dzni?file=src%2Fapp%2Fmycomponent%2Fmycomponent.html,src%2Fapp%2Fmycomponent%2Fmycomponent.ts,src%2Fmain.ts&view=editor

In the example above, the rule fails with this error:

```
/home/projects/angular-eslint-repros-rft8dzni/src/main.ts
  10:41  error  Avoid using the non-null assertion operator (!) in templates. This bypasses type safety and can lead to runtime errors  @angular-eslint/template/no-non-null-assertion
```

However, if I remove the non-null assertion operator from that location, the project fails to build:

```sh
❯ ng build
Application bundle generation failed. [10.726 seconds] - 2026-03-06T23:14:21.756Z

✘ [ERROR] TS2322: Type 'string | undefined' is not assignable to type 'string'.
  Type 'undefined' is not assignable to type 'string'. [plugin angular-compiler]

    src/main.ts:10:24:
      10 │       <app-mycomponent [myStringInput]="maybeUndefined()" />
         ╵                         ~~~~~~~~~~~~~
```

Additionally, the error happens even if the block in question is already guarded against nullness.

So this is a request to not error, either in cases where the assertion is needed to build, or at least in cases where the template already guards against it.

**Versions**

```sh
# Please run `npx ng version` in your project and paste the full output here:

❯ npx ng version

     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI       : 21.2.1
Angular           : 21.2.1
Node.js           : 20.19.1
Package Manager   : npm 10.8.2
Operating System  : linux x64

┌───────────────────────────┬───────────────────┬───────────────────┐
│ Package                   │ Installed Version │ Requested Version │
├───────────────────────────┼───────────────────┼───────────────────┤
│ @angular/animations       │ 21.2.1            │ ^21.2.0           │
│ @angular/build            │ 21.2.1            │ ^21.2.0           │
│ @angular/cdk              │ 21.2.1            │ ^21.2.0           │
│ @angular/cli              │ 21.2.1            │ ^21.2.0           │
│ @angular/common           │ 21.2.1            │ ^21.2.0           │
│ @angular/compiler         │ 21.2.1            │ ^21.2.0           │
│ @angular/compiler-cli     │ 21.2.1            │ ^21.2.1           │
│ @angular/core             │ 21.2.1            │ ^21.2.0           │
│ @angular/forms            │ 21.2.1            │ ^21.2.1           │
│ @angular/platform-browser │ 21.2.1            │ ^21.2.0           │
│ @angular/router           │ 21.2.1            │ ^21.2.1           │
│ rxjs                      │ 7.8.2             │ ~7.8.2            │
│ typescript                │ 5.9.3             │ ~5.9.3            │
│ zone.js                   │ 0.16.1            │ ^0.16.1           │
└───────────────────────────┴───────────────────┴───────────────────┘
```

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest supported version of the packages and checked my `ng version` output per the instructions given here.

**Comment by undsoft** [comment]
Hey.

It's an interesting case.
Generally speaking the rule behaves in line [with how non-null-assertion works](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=GYVwdgxgLglg9mABAWwJ4GUYHMwEMA2AFAJQBciAzlAE4xhaIA%2Bi4AJgKbB3uuIDeAKESJq7KCGpIA5FIDcAgL4CB%2BMYgBuBEO3JVa9eTGCJCaTDgIli-IRq3tEAXhQZseIsQCE8pQKA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1s6e4XloBDZMkTR8lDk3RQx0DtEjgwAXxAqgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false) in Typescript eslint.

It's you who knows that `maybeUndefined()` is a signal and will return the same value every time when called. 
But for Typescript it's just two separate function calls each of which may return either a string or an undefined.

Doing something like this will help Typescript understand that it's the same value:

```html
@let myString = maybeUndefined();
@if (myString) {
  <app-mycomponent [myStringInput]="myString" />
}
```

However, I do agree that it's annoying, although I don't know who can do something about it. 
It seems unlikely that Angular compiler will start treating signal calls in a special way compared to the other Typescript functions.
And it's not for the linter to make an exception because the rule is built to be generic enough.

**Comment by rdamazio** [comment]
Indeed, your suggestion is a good workaround - thanks for that.

> And it's not for the linter to make an exception because the rule is built to be generic enough.

Well, IMHO since it's an *Angular* linter, treating standard Angular primitives differently wouldn't be that big a stretch, but I understand your point.

**Comment by JamesHenry** [comment]
@undsoft's feedback seems correct to me here, so I'll close this one
